pax/plugins/signal_processing/DesaturatePulses.py

```diff
@@ -27,71 +27,117 @@ def transform_event(self, event):
 
             # Where is the current pulse saturated?
             saturated = pulse.raw_data <= 0            # Boolean array, True if sample is saturated
-            _where_saturated = np.where(saturated)[0]
-            try:
-                first_saturated = _where_saturated.min()
-                last_saturated = _where_saturated.max()
-            except (ValueError, RuntimeError, TypeError, NameError):
-                continue
+            _where_saturated_all = np.where(saturated)[0]
+
+            # Split saturation if there is long enough non-saturated samples in between
+            _where_saturated_diff = np.diff(_where_saturated_all, n=1)
+            _where_saturated_diff = np.where(_where_saturated_diff > self.config['reference_region_samples'])[0]
+            _where_saturated_list = np.split(_where_saturated_all, _where_saturated_diff+1)
 
-            # Select a reference region just before the start of the saturated region
-            reference_slice = slice(max(0, first_saturated - self.config['reference_region_samples']),
-                                    first_saturated)
             # Find all pulses in TPC channels that overlap with the saturated & reference region
             other_pulses = [p for i, p in enumerate(event.pulses)
-                            if p.left < last_saturated + pulse.left and p.right > pulse.left and
+                            if p.left < pulse.right and p.right > pulse.left and
                             not is_saturated[i] and
-                            p.channel in tpc_channels]
+                            p.channel in tpc_channels and
+                            p.channel not in self.config['large_after_pulsing_channels']]
```

Comment:
What if only those large after pulsing channels are not saturated? I guess it's not very likely?

Reply:
Doesn't seem likely, but not sure what will happen if other_pulses happen to be empty.